File: thinking-categorically/domain-and-codomain/03-partial-functions/partial-functions.cs
// Partial function using nullable return type
static int? ParseInteger(string input)
{
    if (int.TryParse(input, out int result))
    {
        return result;
    }

    return null;  // Explicit representation of failure
}

// Usage with null checking
string userInput = "42";
int? parsed = ParseInteger(userInput);
if (parsed.HasValue)
{
    Console.WriteLine($"Result: {parsed.Value * 2}");
}
else
{
    Console.WriteLine("Invalid input");
}

// Domain: string (all strings)
// Codomain: int? (nullable integer)

File: thinking-categorically/domain-and-codomain/03-partial-functions/partial-functions.ts
// Partial function - might fail
function parseInteger(input: string): number | undefined {
  const result = parseInt(input);
  return isNaN(result) ? undefined : result;
}

// Usage with explicit null checking
const userInput = "42";
const parsed = parseInteger(userInput);
if (parsed !== undefined) {
  console.log(`Result: ${parsed * 2}`);
} else {
  console.log("Invalid input");
}

// Domain: string (all strings)
// Codomain: number | undefined (number or absence of value)